src/resolver.rs

```diff
@@ -16,23 +15,47 @@ pub struct ResolvesServerCertAcme {
 #[derive(Debug)]
 struct Inner {
     cert: Option<Arc<CertifiedKey>>,
-    auth_keys: BTreeMap<String, Arc<CertifiedKey>>,
+    challenge_data: Option<ChallengeData>,
+}
+
+#[derive(Debug)]
+enum ChallengeData {
+    TlsAlpn01 { sni: String, cert: Arc<CertifiedKey> },
+    Http01 { token: String, key_auth: Arc<Vec<u8>> },
 }
 
 impl ResolvesServerCertAcme {
     pub(crate) fn new() -> Arc<Self> {
         Arc::new(Self {
             inner: Mutex::new(Inner {
                 cert: None,
-                auth_keys: Default::default(),
+                challenge_data: None,
             }),
         })
     }
     pub(crate) fn set_cert(&self, cert: Arc<CertifiedKey>) {
         self.inner.lock().unwrap().cert = Some(cert);
     }
-    pub(crate) fn set_auth_key(&self, domain: String, cert: Arc<CertifiedKey>) {
-        self.inner.lock().unwrap().auth_keys.insert(domain, cert);
+    pub(crate) fn set_tls_alpn_01_challenge_data(&self, domain: String, cert: Arc<CertifiedKey>) {
+        self.inner.lock().unwrap().challenge_data = Some(ChallengeData::TlsAlpn01 { sni: domain, cert });
+    }
+    pub(crate) fn set_http_01_challenge_data(&self, token: String, key_auth: Arc<Vec<u8>>) {
+        self.inner.lock().unwrap().challenge_data = Some(ChallengeData::Http01 { token, key_auth })
+    }
+    pub(crate) fn clear_challenge_data(&self) {
+        self.inner.lock().unwrap().challenge_data = None;
+    }
+    pub fn get_key_auth(&self, challenge_token: &String) -> Option<Arc<Vec<u8>>> {
+        match &self.inner.lock().unwrap().challenge_data {
+            Some(ChallengeData::Http01 { token, key_auth }) => {
```

Comment:
Validating the token is the expected challenge token

Reply:
:+1: , yes, that's necessary. maybe use `==` instead of `.eq()`

Reply:
Ah yes. Java day job strikes again.